src/java.base/share/classes/sun/security/util/CryptoAlgorithmConstraints.java

```diff
@@ -0,0 +1,145 @@
+
+/**
+ * Algorithm constraints for disabled crypto algorithms property
+ *
+ * See the "jdk.crypto.disabledAlgorithms" specification in java.security
+ * for the syntax of the disabled algorithm string.
+ */
+public class CryptoAlgorithmConstraints extends AbstractAlgorithmConstraints {
+    private static final Debug debug = Debug.getInstance("jca");
+
+    private static class CryptoHolder {
+        static final CryptoAlgorithmConstraints CONSTRAINTS =
+                new CryptoAlgorithmConstraints("jdk.crypto.disabledAlgorithms");
+    }
+
+    private static void debug(String msg) {
+        if (debug != null) {
+            debug.println("CryptoAlgoConstraints: ", msg);
+        }
+    }
+
+    public static final boolean permits(String service, String algo) {
+        return CryptoHolder.CONSTRAINTS.permits(null, service + "." + algo, null);
+    }
+
+    private final Set<String> disabledServices; // syntax is <service>.<algo>
+    private volatile SoftReference<Map<String, Boolean>> cacheRef =
+            new SoftReference<>(null);
+
+    /**
+     * Initialize algorithm constraints with the specified security property.
+     *
+     * @param propertyName the security property name that define the disabled
+     *        algorithm constraints
+     */
+    CryptoAlgorithmConstraints(String propertyName) {
+        super(null);
+        disabledServices = getAlgorithms(propertyName);
+        debug("Before " + Arrays.deepToString(disabledServices.toArray()));
+        for (String dk : disabledServices) {
+            int idx = dk.indexOf(".");
+            if (idx == -1) {
+                debug("Remove invalid entry: " + dk);
+                disabledServices.remove(dk);
+                continue;
+            }
+            String service = dk.substring(0, idx);
+            String algo = dk.substring(idx + 1);
+            KnownOIDs oid = KnownOIDs.findMatch(algo);
+            if (oid != null) {
+                debug("Add oid: " + oid.value());
+                disabledServices.add(service + "." + oid.value());
+                debug("Add oid stdName: " + oid.stdName());
+                disabledServices.add(service + "." + oid.stdName());
+                for (String a : oid.aliases()) {
+                    debug("Add oid alias: " + a);
+                    disabledServices.add(service + "." + a);
+                }
+            }
+        }
+        debug("After " + Arrays.deepToString(disabledServices.toArray()));
+    }
+
+    /*
+     * This checks if the specified service descriptor is in the
+     * disabledServices Set. If found, this method return false.
+     */
+    @Override
+    public final boolean permits(Set<CryptoPrimitive> notUsed1,
+            String serviceDesc, AlgorithmParameters notUsed2) {
+        if (serviceDesc == null || serviceDesc.isEmpty()) {
+            throw new IllegalArgumentException("No algorithm name specified");
+        }
+
+        return cachedCheckAlgorithm(serviceDesc);
```

Comment:
Is this doing a case-insensitive check? I was trying to figure that out, but I think "md2" should still match "MD2" for example.

Reply:
Yes, the tests do make sure the check is case-insensitive.